### PR TITLE
Initial setup

### PR DESCRIPTION
**Pull request by rchl**
@bitsper2nd @TheSecEng @jfcherng

### CODE REVIEW

File: plugin.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            data=None,
+            headers={
+                'User-Agent': USER_AGENT
+            }
+        )
+        marksman_path = cls.marksman_path()
+        with urllib.request.urlopen(req) as fp:
+            os.makedirs(os.path.dirname(marksman_path), exist_ok=True)
+            with open(marksman_path, 'wb') as f:
+                f.write(fp.read())
```
Comment (jfcherng) [inline plugin.py:81]:
```suggestion
        marksman_path = cls.marksman_path()
        os.makedirs(os.path.dirname(marksman_path), exist_ok=True)
        bin_url = MARKSMAN_RELEASES_BASE.format(tag=cls.server_version(), platform=marksman_binary())
        req = urllib.request.Request(
            bin_url,
            data=None,
            headers={
                'User-Agent': USER_AGENT
            }
        )
        with urllib.request.urlopen(req) as fp:
            with open(marksman_path, 'wb') as f:
                f.write(fp.read())
```

Comment (jfcherng) [reply plugin.py:81]: hmm... having no idea why the diff looks so broken. I basically move `os.makedirs(os.path.dirname(marksman_path), exist_ok=True)` to the beginning to replace `os.makedirs(cls.basedir(), exist_ok=True)`.

File: plugin.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            with open(os.path.join(cls.basedir(), 'VERSION'), 'r') as fp:
+                return fp.read()
+        except Exception:
+            return None
+
+    @classmethod
+    def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
+                  workspace_folders: List[WorkspaceFolder], configuration: ClientConfig) -> Optional[str]:
+        if marksman_binary() is None:
+            return 'Platform "{}" is not supported'.format(sublime.platform())
```
Comment (rchl) [inline plugin.py:60]: Hmm, this runs after `install_or_update` so might have to move it back to `needs_update_or_installation`.

File: plugin.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+def marksman_binary() -> Optional[str]:
+    platform = sublime.platform()
+    if platform == 'osx':
+        return 'marksman-macos'
+    if platform == 'windows':
+        return 'marksman.exe'
+    if platform == 'linux':
+        return 'marksman-linux'
+    return None
```
Comment (jfcherng) [inline plugin.py:22]:
```suggestion
def marksman_binary() -> Optional[str]:
    platform_arch = "{}_{}".format(sublime.platform(), sublime.arch())
    if platform_arch == 'osx_x64':
        return 'marksman-macos'
    if platform_arch == 'windows_x64':
        return 'marksman.exe'
    if platform_arch == 'linux_x64':
        return 'marksman-linux'
    return None
```

Those pre-builts are x64 https://github.com/artempyanykh/marksman/runs/8049673045

image
image
image

Comment (rchl) [reply plugin.py:22]: It still work on arm Mac through emulation. Better than nothing.

Comment (jfcherng) [reply plugin.py:22]: suggestion revised.

Comment (rchl) [reply plugin.py:22]: But makes sense to do for other platforms

Comment (jfcherng) [reply plugin.py:22]: Suggestion revised.

**Pull request merged**